SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -260,36 +273,55 @@ void USpatialSender::SendComponentUpdates(UObject* Object, const FClassInfo& Inf
+	USpatialLatencyTracer* Tracer = nullptr;
+#if TRACE_LIB_ACTIVE
+	Tracer = USpatialLatencyTracer::GetTracer(Object);
+#endif
+	ComponentFactory UpdateFactory(Channel->GetInterestDirty(), NetDriver, Tracer);
+
+	TArray<TraceKey>* TraceKeysPtr = nullptr;
+#if TRACE_LIB_ACTIVE
+	TArray<TraceKey> TraceKeys;
+	TraceKeysPtr = &TraceKeys;
+#endif
 
-	TArray<Worker_ComponentUpdate> ComponentUpdates = UpdateFactory.CreateComponentUpdates(Object, Info, EntityId, RepChanges, HandoverChanges);
+	TArray<Worker_ComponentUpdate> ComponentUpdates = UpdateFactory.CreateComponentUpdates(Object, Info, EntityId, RepChanges, HandoverChanges, TraceKeysPtr);
 
-	for (Worker_ComponentUpdate& Update : ComponentUpdates)
+	for(int i = 0; i < ComponentUpdates.Num(); i++)
 	{
+		Worker_ComponentUpdate& Update = ComponentUpdates[i];
+		TraceKey LatencyKey = TraceKeysPtr ? (*TraceKeysPtr)[i] : USpatialLatencyTracer::InvalidTraceKey;
 		if (!NetDriver->StaticComponentView->HasAuthority(EntityId, Update.component_id))
 		{
 			UE_LOG(LogSpatialSender, Verbose, TEXT("Trying to send component update but don't have authority! Update will be queued and sent when authority gained. Component Id: %d, entity: %lld"), Update.component_id, EntityId);
 
 			// This is a temporary fix. A task to improve this has been created: UNR-955
 			// It may be the case that upon resolving a component, we do not have authority to send the update. In this case, we queue the update, to send upon receiving authority.
 			// Note: This will break in a multi-worker context, if we try to create an entity that we don't intend to have authority over. For this reason, this fix is only temporary.
-			TArray<Worker_ComponentUpdate>& UpdatesQueuedUntilAuthority = UpdatesQueuedUntilAuthorityMap.FindOrAdd(EntityId);
-			UpdatesQueuedUntilAuthority.Add(Update);
+			FQueuedUpdate& UpdatesQueuedUntilAuthority = UpdatesQueuedUntilAuthorityMap.FindOrAdd(EntityId);
+			UpdatesQueuedUntilAuthority.ComponentUpdates.Add(Update);
+			UpdatesQueuedUntilAuthority.LatencyKeys.Add(LatencyKey);
 			continue;
 		}
 
-		Connection->SendComponentUpdate(EntityId, &Update);
+		Connection->SendComponentUpdate(EntityId, &Update, LatencyKey);
 	}
 }
 
 // Apply (and clean up) any updates queued, due to being sent previously when they didn't have authority.
 void USpatialSender::ProcessUpdatesQueuedUntilAuthority(Worker_EntityId EntityId)
 {
-	if (TArray<Worker_ComponentUpdate>* UpdatesQueuedUntilAuthority = UpdatesQueuedUntilAuthorityMap.Find(EntityId))
+	if (FQueuedUpdate* UpdatesQueuedUntilAuthority = UpdatesQueuedUntilAuthorityMap.Find(EntityId))
 	{
-		for (Worker_ComponentUpdate& Update : *UpdatesQueuedUntilAuthority)
+		for (int i = 0; i < UpdatesQueuedUntilAuthority->ComponentUpdates.Num(); i++)
 		{
-			Connection->SendComponentUpdate(EntityId, &Update);
+			TArray<Worker_ComponentUpdate>& Components = UpdatesQueuedUntilAuthority->ComponentUpdates;
+			TArray<TraceKey>& LatencyKeys = UpdatesQueuedUntilAuthority->LatencyKeys;
```

Comment:
Ideally the additional processing when LIB_TRACE is false is minimal. Can we cull this somewhat?

Reply:
Yeah, can wrap these in that macro (and the definition)